experimental/virtio/src/queue.rs

```diff
@@ -0,0 +1,606 @@
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-380006>.
+#[repr(C, align(2))]
+struct AvailRing<const QUEUE_SIZE: usize> {
+    /// Driver-specific flags for the queue.
+    flags: RingFlags,
+    /// The next index that will be used in the ring (modulo `QUEUE_SIZE`).
+    ///
+    /// Starts at 0 and always increments. Must never be decremented.
+    idx: u16,
+    /// The ring-buffer containing indices of the heads of available descriptor chains.
+    ring: [u16; QUEUE_SIZE],
+}
+
+impl<const QUEUE_SIZE: usize> Default for AvailRing<QUEUE_SIZE> {
+    fn default() -> Self {
+        Self {
+            // We implement all drivers via polling, so don't want notifications.
+            flags: RingFlags::NO_NOTIFY,
+            idx: 0,
+            ring: [0; QUEUE_SIZE],
+        }
+    }
+}
+
+/// The ring buffer that indicates which available descriptors have been consumed by the device.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-430008>.
+#[repr(C, align(4))]
+struct UsedRing<const QUEUE_SIZE: usize> {
+    /// Device-specific flags for the queue.
+    flags: RingFlags,
+    /// The next index that will be used in the ring (modulo `QUEUE_SIZE`).
+    ///
+    /// Starts at 0 and always increments. Must never be decremented.
+    idx: u16,
+    /// The ring-buffer containing the used elements.
+    ring: [UsedElem; QUEUE_SIZE],
+}
+
+impl<const QUEUE_SIZE: usize> Default for UsedRing<QUEUE_SIZE> {
+    fn default() -> Self {
+        Self {
+            flags: RingFlags::empty(),
+            idx: 0,
+            ring: [(); QUEUE_SIZE].map(|_| UsedElem::default()),
+        }
+    }
+}
+
+/// An element indicating a used descriptor chain.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-430008>.
+#[repr(C)]
+#[derive(Clone, Copy, Default)]
+struct UsedElem {
+    /// The index of the head of the used descriptor chain.
+    id: u32,
+    /// Total length of the bytes that was written to the used descriptor chain (only used for
+    /// device-write-only descriptors).
+    len: u32,
+}
+
+/// A split virtqueue implementation.
+///
+/// See <https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-240006>.
+#[repr(C, align(64))]
+struct VirtQueue<const QUEUE_SIZE: usize> {
+    /// The descriptor table.
+    desc: [Desc; QUEUE_SIZE],
+    /// The available ring, which is controlled by the driver.
+    avail: AvailRing<QUEUE_SIZE>,
+    /// The used ring, which is controlled by the device.
+    used: UsedRing<QUEUE_SIZE>,
+}
+
+/// The intended use for the queue.
+#[derive(PartialEq, Eq)]
+enum QueueType {
+    DeviceWriteOnly,
+    DriverWriteOnly,
+}
+
+/// A container for a boxed virtqueue and its associated buffers.
+pub struct Queue<const QUEUE_SIZE: usize, const BUFFER_SIZE: usize> {
```

Comment:
Given that github is already complaining about the size of this file, maybe split the file in two: one that contains all the virtio data structures from the spec, and one that just implements the `Queue`.